R/utils-ui.R

```diff
@@ -1,12 +1,264 @@
+  if (cli::num_ansi_colors() > 1) {
+    x
+  } else {
+    paste0(quote, x, quote)
+  }
+}
+
+# silence -----------------------------------------------------------------
+#' Suppress usethis's messaging
+#'
+#' Execute a bit of code without usethis's normal messaging.
+#'
+#' @param code Code to execute with usual UI output silenced.
+#'
+#' @returns Whatever `code` returns.
+#' @export
+#' @examples
+#' # compare the messaging you see from this:
+#' browse_github("usethis")
+#' # vs. this:
+#' ui_silence(
+#'   browse_github("usethis")
+#' )
+ui_silence <- function(code) {
+  withr::with_options(list(usethis.quiet = TRUE), code)
+}
+
+is_quiet <- function() {
+  isTRUE(getOption("usethis.quiet", default = FALSE))
+}
+
+# bullets, helpers, and friends ------------------------------------------------
+ui_bullets <- function(text, .envir = parent.frame()) {
+  if (is_quiet()) {
+    return(invisible())
+  }
+  cli::cli_div(theme = usethis_theme())
+  cli::cli_bullets(text, .envir = .envir)
+}
+
+ui_path_impl <- function(x, base = NULL) {
+  is_directory <- is_dir(x) | grepl("/$", x)
+  if (is.null(base)) {
+    x <- proj_rel_path(x)
+  } else if (!identical(base, NA)) {
+    x <- path_rel(x, base)
+  }
+
+  # rationalize trailing slashes
+  x <- path_tidy(x)
+  x[is_directory] <- paste0(x[is_directory], "/")
+
+  unclass(x)
+}
+
+# shorter form for compactness, because this is typical usage:
+# ui_bullets("blah blah {.path {pth(some_path)}}")
+pth <- ui_path_impl
+
+ui_code_snippet <- function(x,
+                            copy = rlang::is_interactive(),
+                            language = c("R", ""),
+                            interpolate = TRUE,
+                            .envir = parent.frame()) {
+  language <- arg_match(language)
+
+  indent <- function(x, first = "  ", indent = first) {
+    x <- gsub("\n", paste0("\n", indent), x)
+    paste0(first, x)
+  }
+
+  x <- glue_collapse(x, "\n")
+  if (interpolate) {
+    x <- glue(x, .envir = .envir)
+    # what about literal `{` or `}`?
+    # use `interpolate = FALSE`, if appropriate
+    # double them, i.e. `{{` or `}}`
+    # open issue/PR about adding `.open` and `.close`
+  }
+
+  if (!is_quiet()) {
+    # the inclusion of `.envir = .envir` leads to test failure
+    # I'm consulting with Gabor on this
+    # leaving it out seems fine for my use case
+    # cli::cli_code(indent(x), language = language, .envir = .envir)
+    cli::cli_code(indent(x), language = language)
+  }
+
+  if (copy && clipr::clipr_available()) {
+    x_no_ansi <- cli::ansi_strip(x)
+    clipr::write_clip(x_no_ansi)
+    style_subtle <- cli::combine_ansi_styles(
+      cli::make_ansi_style("grey"),
+      cli::style_italic
+    )
+    ui_bullets(c(" " = style_subtle("[Copied to clipboard]")))
+  }
+
+  invisible(x)
+}
+
+# inspired by gargle::gargle_map_cli() and gargle::bulletize()
+usethis_map_cli <- function(x, ...) UseMethod("usethis_map_cli")
```

Comment:
I have several functions related to the need to programmatically make bullets or to build up bullets in stages, e.g. preinterpolate data. I suspect this could be refactored in the future, but what I have produces the output that I want.